plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,331 @@
+   */
+  override fun addStatusObserver(viewportStatusObserver: ViewportStatusObserver) {
+    registeredStatusObservers.add(viewportStatusObserver)
+  }
+
+  /**
+   * Removes [ViewportStatusObserver].
+   */
+  override fun removeStatusObserver(viewportStatusObserver: ViewportStatusObserver) {
+    registeredStatusObservers.remove(viewportStatusObserver)
+  }
+
+  // Convenient methods to create the in-stock [ViewportState] and [ViewportTransition].
+
+  /**
+   * Create a [FollowingViewportState] instance with provided [FollowingViewportStateOptions].
+   *
+   * @param options The desired [FollowingViewportStateOptions]
+   */
+  override fun makeFollowingViewportState(options: FollowingViewportStateOptions): FollowingViewportState {
+    return FollowingViewportStateImpl(delegateProvider, options)
+  }
+
+  /**
+   * Create an [OverviewViewportState] instance with provided [OverviewViewportStateOptions].
+   *
+   * @param options The desired [OverviewViewportStateOptions]
+   */
+  override fun makeOverviewViewportState(options: OverviewViewportStateOptions): OverviewViewportState {
+    return OverviewViewportStateImpl(delegateProvider, options)
+  }
+
+  /**
+   * Create a default [ViewportTransition] instance with provided [DefaultViewportTransitionOptions].
+   *
+   * @param options The desired [DefaultViewportTransitionOptions]
+   */
+  override fun makeDefaultTransition(options: DefaultViewportTransitionOptions): DefaultViewportTransition {
+    return DefaultViewportTransitionImpl(delegateProvider, options)
+  }
+
+  /**
+   * Create a [ViewportTransition] instance that transition to the target [ViewportState] immediately.
+   */
+  override fun makeImmediateViewportTransition(): ViewportTransition {
+    return ImmediateViewportTransition(delegateProvider)
+  }
+
+  /**
+   * Provides all map delegate instances.
+   */
+  override fun onDelegateProvider(delegateProvider: MapDelegateProvider) {
+    this.delegateProvider = delegateProvider
+    this.cameraPlugin = delegateProvider.mapPluginProviderDelegate.camera
+    cameraPlugin.addCameraAnimationsLifecycleListener(cameraAnimationsLifecycleListener)
+    this.defaultTransition = DefaultViewportTransitionImpl(
+      delegateProvider,
+      DefaultViewportTransitionOptions.Builder().build()
+    )
+  }
+
+  /**
+   * Called when the map is destroyed. Should be used to cleanup plugin resources for that map.
+   */
+  override fun cleanup() {
+    updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC)
+    cameraPlugin.removeCameraAnimationsLifecycleListener(cameraAnimationsLifecycleListener)
+  }
+
+  internal companion object {
+    /**
+     * Constant used to recognize the owner of transitions initiated by the [ViewportPluginImpl].
+     *
+     * @see CameraAnimator.owner
+     */
+    internal const val VIEWPORT_CAMERA_OWNER = "VIEWPORT_CAMERA_OWNER"
+  }
+}
+
+internal fun ViewportStatus.getCurrentOrTargetState(): ViewportState? =
```

Comment:
I'm seeing it used across this PR for several times - perhaps expose it and re-use once?

Reply:
It was specifically removed from previous code reviews, because this extension seems unclear or have very limited value on specific use cases. User could implement this logic easily as well.